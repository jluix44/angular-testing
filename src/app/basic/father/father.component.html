<h1>Padre</h1>

<button class="btn btn-primary" (click)="onSetClient('Jose Luis')">
    Cargar clinte
</button>

<br><br>

<code class="mt-2">{{ client | json }}</code>

<hr>

<app-son [client]="client" (onDeleteClient)="client = undefined" (onClientUpdated)="client = $event "></app-son>