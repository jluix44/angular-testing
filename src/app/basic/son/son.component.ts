import { Component, OnInit, Input, Output, EventEmitter } from '@angular/core';
import { Client } from '../interfaces';

@Component({
    selector: 'app-son',
    templateUrl: './son.component.html',
    styleUrls: ['./son.component.css']
})
export class SonComponent implements OnInit {
    
    @Input() client?: Client
    @Output() onDeleteClient = new EventEmitter()
    @Output() onClientUpdated = new EventEmitter<Client>()
    
    constructor() { }
    
    ngOnInit(): void {
    }
    

    onDelete(){
        this.client = undefined
        this.onDeleteClient.emit()
    }

    onChange(id: number){
        if(!this.client) return

        //this.client.id = id // no usar asi porque se actualizar el obj en father and son

        this.client = { ...this.client, id}
        this.onClientUpdated.emit({...this.client})
    }
}
