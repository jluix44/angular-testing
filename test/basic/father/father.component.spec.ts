import { ComponentFixture, TestBed } from '@angular/core/testing';
import { SonComponent } from '../../../src/app/basic/son/son.component';

import { FatherComponent } from '../../../src/app/basic/father/father.component';
import { By } from '@angular/platform-browser';

describe('FatherComponent', () => {
    let component: FatherComponent;
    let fixture: ComponentFixture<FatherComponent>;
    let compiled: HTMLElement;
    
    beforeEach(async () => {
        await TestBed.configureTestingModule({
            declarations: [ FatherComponent, SonComponent ]
        })
        .compileComponents();
    });
    
    beforeEach(() => {
        fixture = TestBed.createComponent(FatherComponent);
        component = fixture.componentInstance;
        fixture.detectChanges();
        compiled = fixture.nativeElement;
    });
    
    test('should create', () => {
        expect(component).toBeTruthy();
    });


    test('debe hacer match con el snapshop', () => {
        expect(compiled).toMatchSnapshot()
    })


    test('debe establecer el cliente con el nombre indicado', () => {
        component.onSetClient('Pedro')
        fixture.detectChanges()


        const codeTag = compiled.querySelector('.mt-2')
        console.log(codeTag?.textContent)

        expect(codeTag?.textContent).toContain('"name"')
        expect(codeTag?.textContent).toContain('"Pedro"')

    })


    test('debe borrar al cliente si se emite onDeleteClient (hijo)', () => {
        component.client = {id: 1, name: 'Eduardo'}
        fixture.detectChanges()

        const sonDebugElement = fixture.debugElement.query(By.directive(SonComponent))
        const sonComponent: SonComponent = sonDebugElement.componentInstance

        sonComponent.onDeleteClient.emit()
        expect(component.client).toBe(undefined) //toBe para evaluar dos datos primitivos stringm, number,

    })



    test('debe actualizar el cliente onClientUpdated', () => {
        component.client = {id: 1, name: 'Eduardo'}
        fixture.detectChanges()

        const sonDebugElement = fixture.debugElement.query(By.directive(SonComponent))
        const sonComponent: SonComponent = sonDebugElement.componentInstance

        sonComponent.onClientUpdated.emit({id: 10, name: 'Pedro'})
        expect(component.client).toEqual({id: 10, name: 'Pedro'}) //toEqual para evaluar cuando un obj == obj

    })

});
