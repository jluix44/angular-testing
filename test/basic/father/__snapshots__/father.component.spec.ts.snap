// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherComponent debe hacer match con el snapshop 1`] = `
<div
  id="root1"
>
  <h1>
    Padre
  </h1>
  <button
    class="btn btn-primary"
  >
     Cargar clinte

  </button>
  <br />
  <br />
  <code
    class="mt-2"
  >
    
  </code>
  <hr />
  <app-son>
    <h5>
      Hijo
    </h5>
    <code>
      
    </code>
    <br />
    <br />
    
    
  </app-son>
</div>
`;
